Use BrowseSync for supporting livereload.

jessy1092 opened the pull request:
And add some npm script:
- "start": "grunt browserify && electron main.js",
- "livereload": "electron main.js --livereload",
- "server": "grunt",
- "build": "grunt build",

jessy1092 committed:
Use BrowserSync for supporting livereload.

Signed-off-by: Lee <jessy1092@gmail.com>

kidwm commented:
livereload 在這裡沒有很大的用處吧，而且不要把不相干的修改混進來…

kidwm commented:
@jessy1092 FYI: https://github.com/Quramy/electron-connect

jessy1092 commented:
以後 plasmon 要加各種功能，也是在 app.js 上開發吧? 像多個分頁、紀錄站台等等。
我是想說方便開發就先加了，如果覺得現在沒必要也可以忽略。  XD

改動到 main.js 是因為 livereload 後，telnet 連線還在，會導致 error 
(之後如果要多個分頁，也會有這個問題。) 
故去撿查連線，假如存在的話，就給舊的資料，維持在上一個畫面。

我加 preprocess 了，在 build 時，應該不會把 livereload.js 加進去。

kidwm commented:
@jessy1092 preprocess 應該是不需要的，可以用 preload 來試試看 https://github.com/atom/electron/blob/master/docs/api/browser-window.md

jessy1092 commented:
@kidwm  試了無法使用 preload 會發生 以下錯誤，我猜跟 LiveReload 的機制有關吧。
`LiveReload disabled because it could not find its own <SCRIPT> tag` 

使用 preprocess 是比較快的解決方式。

kidwm commented:
@jessy1092 那換用 browserSync 然後改變 main.js 裡面載入的 url 如何？

jessy1092 commented:
@kidwm OK 剛試出來了~
1. 先開啟 BrowserSync 的 Server
2. `electron main.js --livereload` 會去 load `http://localhost:3000/`

這樣如何?

kidwm commented:
@jessy1092 grunt server 應該可以直接跟 default 合併？

jessy1092 commented:
@kidwm  done

kidwm commented:
@jessy1092 很好，最後就是用了 browserSync 其實就不需要 grunt-contrib-watch 了，你可以把它拿掉，然後把 browserify 的 watch 加回來試試

jessy1092 committed:
Handle existing telnet connection for livereload

After livereload app.js, the telnet connection was existing on main.js.
It would trigger the connection connect again and occur error.

So I check whether connection was existing or not.
If not, return the old telnet data.

Signed-off-by: Lee <jessy1092@gmail.com>

jessy1092 commented:
@kidwm  :+1:  :+1:   原來可以直接這樣~  感謝提醒!  done

kidwm commented:
@jessy1092 Grunt 的執行順序應該要先跑 browserify 才正確吧？

另外 grunt-cli 我記得是要裝在系統而不是專案裡面的？

jessy1092 commented:
@kidwm 
browserSync 會自動監聽新產生的 file。
不過主要原因是: 如果 browserify 在前面，因為 browserify watch 開啟就會 hang 住，不會在執行 browserSync。

grun-cli 主要是給沒有安裝 grunt 的人使用 (像我系統是沒安裝 grunt)，不過使用 npm script 時，會自動去 node_modules 下找 grunt-cli ，讓他可以執行。
這樣開發者就不用管他系統有沒有安裝 grunt 了， 只要 `npm run server` 跟 `npm run livereload` 就可以了，而且專案也不會因為 開發者 系統內的 grunt 版本不同而不能跑。

kidwm commented:
@jessy1092 你 browserify 沒把 watch 加回去當然會 hang 住啊

jessy1092 commented:
@kidwm  不是把 browserify option 裡面的 watch 跟 keepAlive 改成 true 嗎? 我以為 hang 住是因為已經在 watch 的關係。

kidwm commented:
@jessy1092 講錯，是你忘了把 keepAlive 拿掉

kidwm commented:
@jessy1092 我剛又看了一下，好像 browserSync 沒有辦法 keepAlive，所以也只能這樣了

kidwm commented:
@jessy1092 有啦有啦，把 browserSync 的 watchTask 也拿掉就會 keepAlive 了，還以為我記錯了 XD

jessy1092 commented:
@kidwm 我剛也試出來了，就變成 BrowserSync 也是個 watcher 。 XD

kidwm commented:
@jessy1092 我順手就按下了，你再直接改 master 吧 XD

jessy1092 commented:
@kidwm  OK! Thanks.
